packages/exchange/src/exchange/providers/groq.py

```diff
@@ -0,0 +1,97 @@
+import os
+from typing import Any, Dict, List, Tuple, Type
+
+import httpx
+
+from exchange.message import Message
+from exchange.providers.base import Provider, Usage
+from exchange.providers.utils import (
+    messages_to_openai_spec,
+    openai_response_to_message,
+    openai_single_message_context_length_exceeded,
+    raise_for_status,
+    tools_to_openai_spec,
+)
+from exchange.tool import Tool
+from tenacity import retry, wait_fixed, stop_after_attempt
+from exchange.providers.utils import retry_if_status
+
+GROQ_HOST = "https://api.groq.com/openai/"
+
+retry_procedure = retry(
+    wait=wait_fixed(5),
+    stop=stop_after_attempt(5),
+    retry=retry_if_status(codes=[429], above=500),
+    reraise=True,
+)
+
+
+class GroqProvider(Provider):
+    """Provides chat completions for models hosted directly by OpenAI."""
+
+    PROVIDER_NAME = "groq"
+    REQUIRED_ENV_VARS = ["GROQ_API_KEY"]
+    instructions_url = "https://console.groq.com/docs/quickstart"
+
+    def __init__(self, client: httpx.Client) -> None:
+        self.client = client
+
+    @classmethod
+    def from_env(cls: Type["GroqProvider"]) -> "GroqProvider":
+        cls.check_env_vars(cls.instructions_url)
+        url = os.environ.get("GROQ_HOST", GROQ_HOST)
+        key = os.environ.get("GROQ_API_KEY")
+
+        client = httpx.Client(
+            base_url=url + "v1/",
+            headers={"Authorization": "Bearer " + key},
+            timeout=httpx.Timeout(60 * 10),
+        )
+        return cls(client)
+
+    @staticmethod
+    def get_usage(data: dict) -> Usage:
```

Comment:
@lifeizhou-ap this might be a useful thing to promote to a `@classmethod` (not sure i understand when we'd need to use it as a static method `Provider.get_usage({...})`). let's consider making an abstraction to make adding providers super easy?